File: src/RadarScene.cpp
#include "RadarScene.h"
#include <iostream>
#include <iomanip>
#include <sstream>
#include <algorithm>
#include <thread>
#include <chrono>

#if USE_COCOS2DX
// ============================================================================
// Cocos2d-x Radar Scene
// ============================================================================

RadarScene::RadarScene()
    : gameConfig_(GameConfig::getInstance())
    , radarDisplay_(nullptr)
    , gameHUD_(nullptr)
    , score_(0)
    , gameOver_(false)
{
}

cocos2d::Scene* RadarScene::createScene()
{
    return RadarScene::create();
}

RadarScene* RadarScene::create()
{
    auto* ret = new (std::nothrow) RadarScene();
    if (ret && ret->init()) {
        ret->autorelease();
        return ret;
    }
    delete ret;
    return nullptr;
}

bool RadarScene::init()
{
    if (!Scene::init()) return false;

    // Initialize game config
    gameConfig_.setLevel(1);

    // Initialize subsystems
    aircraftGenerator_.init(gameConfig_.getLevel());
    fireControlSystem_.init();

    // Set up IFF error rate from level config
    trackManager_.getIFFSystem().setErrorRate(gameConfig_.getIFFErrorRate());

    // Create display elements
    initRadarDisplay();
    initHUD();
    initInputHandlers();

    // Start game loop
    scheduleUpdate();
    return true;
}

void RadarScene::initRadarDisplay()
{
    auto visibleSize = cocos2d::Director::getInstance()->getVisibleSize();

    // Radar takes up the left/center portion of the screen
    float radarRadius = std::min(visibleSize.width * 0.45f,
                                  visibleSize.height * 0.45f);

    radarDisplay_ = RadarDisplay::create(radarRadius);
    radarDisplay_->setPosition(cocos2d::Vec2(
        visibleSize.width * 0.4f, visibleSize.height * 0.5f));
    radarDisplay_->setTrackManager(&trackManager_);
    radarDisplay_->setFireControlSystem(&fireControlSystem_);
    addChild(radarDisplay_, 1);
}

void RadarScene::initHUD()
{
    auto visibleSize = cocos2d::Director::getInstance()->getVisibleSize();

    gameHUD_ = GameHUD::create();
    gameHUD_->setPosition(cocos2d::Vec2(visibleSize.width * 0.75f,
                                         visibleSize.height * 0.95f));
    gameHUD_->setTrackManager(&trackManager_);
    gameHUD_->setFireControlSystem(&fireControlSystem_);
    gameHUD_->setLevel(gameConfig_.getLevel());
    addChild(gameHUD_, 2);
}

void RadarScene::initFireControl()
{
    fireControlSystem_.init();
}

void RadarScene::initInputHandlers()
{
    // Mouse/touch listener for selecting tracks on radar
    auto listener = cocos2d::EventListenerTouchOneByOne::create();
    listener->onTouchBegan = [this](cocos2d::Touch* touch, cocos2d::Event* event) {
        // Convert touch position to radar-local coordinates
        auto touchPos = touch->getLocation();
        auto localPos = radarDisplay_->convertToNodeSpace(touchPos);

        // Use RadarDisplay's findNearestTrack for accurate hit detection
        int nearestTrackId = radarDisplay_->findNearestTrack(localPos, 20.0f);

        if (nearestTrackId >= 0) {
            onTrackSelected(nearestTrackId);
        } else {
            // Clicked empty space — deselect
            onTrackSelected(-1);
        }
        return true;
    };

    getEventDispatcher()->addEventListenerWithSceneGraphPriority(listener, this);

    // Keyboard listener for fire control commands
    auto keyListener = cocos2d::EventListenerKeyboard::create();
    keyListener->onKeyPressed = [this](cocos2d::EventKeyboard::KeyCode keyCode,
                                        cocos2d::Event* event) {
        switch (keyCode) {
            // Number keys 1-6 for battery selection
            case cocos2d::EventKeyboard::KeyCode::KEY_1:
                onBatteryAssign("PATRIOT-1"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_2:
                onBatteryAssign("PATRIOT-2"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_3:
                onBatteryAssign("PATRIOT-3"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_4:
                onBatteryAssign("HAWK-1"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_5:
                onBatteryAssign("HAWK-2"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_6:
                onBatteryAssign("HAWK-3"); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_F:
                onFireAuthorized(); break;
            case cocos2d::EventKeyboard::KeyCode::KEY_A:
                onAbortEngagement(); break;
            default:
                break;
        }
    };

    getEventDispatcher()->addEventListenerWithSceneGraphPriority(keyListener, this);
}

void RadarScene::update(float dt)
{
    if (gameOver_) return;

    spawnAircraft(dt);
    updateAircraft(dt);
    trackManager_.update(dt);
    fireControlSystem_.update(dt);
    checkEngagements(dt);
    cleanupDestroyedAircraft();
    checkGameOver();
}

void RadarScene::spawnAircraft(float dt)
{
    int currentCount = 0;
    for (const auto& ac : aircraft_) {
        if (ac->isAlive()) currentCount++;
    }

    Aircraft* newAircraft = aircraftGenerator_.trySpawn(dt, currentCount);
    if (newAircraft) {
        int trackId = trackManager_.addTrack(newAircraft);
        aircraft_.push_back(std::unique_ptr<Aircraft>(newAircraft));

        std::string msg = "NEW CONTACT: " + newAircraft->getTrackData().getTrackIdString() +
                         " AZ:" + std::to_string((int)newAircraft->getAzimuth()) +
                         " RNG:" + std::to_string((int)newAircraft->getRange()) + "km";
        if (gameHUD_) gameHUD_->addMessage(msg);
    }
}

void RadarScene::updateAircraft(float dt)
{
    for (auto& ac : aircraft_) {
        if (ac->isAlive()) {
            ac->update(dt);

            if (ac->hasReachedTerritory() && !ac->isFriendly()) {
                onHostilePenetrated(ac.get());
                ac->destroy();
            }
        }
    }
}

void RadarScene::checkEngagements(float dt)
{
    auto results = fireControlSystem_.getRecentResults();
    for (const auto& result : results) {
        auto* aircraft = trackManager_.getAircraftByTrackId(result.trackId);
        std::string msg;

        if (result.result == EngagementResult::HIT) {
            msg = result.batteryDesignation + " -> " +
                  (aircraft ? aircraft->getTrackData().getTrackIdString() : "TK-???") +
                  " ** SPLASH **";
            if (aircraft) {
                if (aircraft->isFriendly()) {
                    onFriendlyDestroyed(aircraft);
                } else {
                    onHostileDestroyed(aircraft);
                }
            }
        } else if (result.result == EngagementResult::MISS) {
            msg = result.batteryDesignation + " -> MISS";
            addScore(GameConstants::SCORE_MISSILE_WASTED);
        }

        if (gameHUD_ && !msg.empty()) {
            gameHUD_->addMessage(msg);
        }
    }
}

void RadarScene::cleanupDestroyedAircraft()
{
    aircraft_.erase(
        std::remove_if(aircraft_.begin(), aircraft_.end(),
            [](const std::unique_ptr<Aircraft>& ac) {
                return !ac->isAlive() ||
                       ac->getRange() > GameConstants::RADAR_MAX_RANGE_KM * 1.1f;
            }),
        aircraft_.end());
}

void RadarScene::checkGameOver()
{
    // Game over if score drops too low
    if (score_ < -2000) {
        gameOver_ = true;
        if (gameHUD_) {
            gameHUD_->addMessage("=== GAME OVER - DEFENSE FAILED ===");
        }
    }
}

void RadarScene::onTrackSelected(int trackId)
{
    radarDisplay_->setSelectedTrack(trackId);
    if (gameHUD_) gameHUD_->setSelectedTrack(trackId);
}

void RadarScene::onBatteryAssign(const std::string& batteryDesignation)
{
    if (radarDisplay_->getSelectedTrack() < 0) return;

    auto* battery = fireControlSystem_.getBattery(batteryDesignation);
    auto* aircraft = trackManager_.getAircraftByTrackId(
        radarDisplay_->getSelectedTrack());

    if (battery && aircraft && battery->canEngage(aircraft)) {
        if (battery->engage(aircraft)) {
            std::string msg = batteryDesignation + " ENGAGING " +
                             aircraft->getTrackData().getTrackIdString();
            if (gameHUD_) gameHUD_->addMessage(msg);
        }
    }
}

void RadarScene::onFireAuthorized() {}
void RadarScene::onAbortEngagement() {}

void RadarScene::addScore(int points)
{
    score_ += points;
    if (gameHUD_) gameHUD_->setScore(score_);
}

void RadarScene::onHostileDestroyed(Aircraft* aircraft)
{
    int score = gameConfig_.getHostileDestroyedScore(aircraft->getType());
    addScore(score);
}

void RadarScene::onFriendlyDestroyed(Aircraft* aircraft)
{
    addScore(GameConstants::SCORE_FRIENDLY_DESTROYED);
    if (gameHUD_) {
        gameHUD_->addMessage("*** WARNING: FRIENDLY FIRE ***");
    }
}

void RadarScene::onHostilePenetrated(Aircraft* aircraft)
{
    addScore(GameConstants::SCORE_HOSTILE_PENETRATED);
    if (gameHUD_) {
        gameHUD_->addMessage("ALERT: HOSTILE PENETRATED DEFENSE ZONE");
    }
}

#else
// ============================================================================
// Stub Radar Scene (no cocos2d-x) — Console-based simulation
// ============================================================================

RadarScene* RadarScene::create()
{
    return new RadarScene();
}

bool RadarScene::init()
{
    GameConfig::getInstance().setLevel(1);
    aircraftGenerator_.init(1);
    fireControlSystem_.init();
    trackManager_.getIFFSystem().setErrorRate(0.0f);

    std::cout << "[RadarScene] Systems initialized." << std::endl;
    std::cout << "  - Radar: Raytheon AN/TPS-43E long-range surveillance" << std::endl;
    std::cout << "  - Range: " << (int)GameConstants::RADAR_MAX_RANGE_NM
              << " NM (" << (int)GameConstants::RADAR_MAX_RANGE_KM << " km)" << std::endl;
    std::cout << "  - 3x Patriot Missile Batteries (MPMB) @ 160 km max" << std::endl;
    std::cout << "  - 3x Hawk SAM Batteries (HSAMB) @ 45 km max" << std::endl;
    std::cout << "  - Sweep rate: " << GameConstants::RADAR_SWEEP_RATE_RPM << " RPM" << std::endl;
    std::cout << std::endl;

    return true;
}

void RadarScene::update(float dt)
{
    spawnAircraft(dt);
    updateAircraft(dt);
    trackManager_.update(dt);
    fireControlSystem_.update(dt);
    cleanupDestroyedAircraft();
}

void RadarScene::spawnAircraft(float dt)
{
    int currentCount = 0;
    for (const auto& ac : aircraft_) {
        if (ac->isAlive()) currentCount++;
    }

    Aircraft* newAircraft = aircraftGenerator_.trySpawn(dt, currentCount);
    if (newAircraft) {
        int trackId = trackManager_.addTrack(newAircraft);
        aircraft_.push_back(std::unique_ptr<Aircraft>(newAircraft));

        auto data = newAircraft->getTrackData();
        std::cout << "[NEW CONTACT] " << data.getTrackIdString()
                  << " | Type: " << newAircraft->getTypeName()
                  << " | AZM: " << std::setw(3) << (int)data.azimuth << "\xC2\xB0"
                  << " | RNG: " << std::setw(3) << (int)(data.range * GameConstants::KM_TO_NM)
                  << " NM"
                  << " | ALT: " << data.getAltitudeString()
                  << " | SPD: " << (int)data.speed << " kts"
                  << " | IFF: " << data.getClassificationString()
                  << std::endl;
    }
}

void RadarScene::updateAircraft(float dt)
{
    for (auto& ac : aircraft_) {
        if (ac->isAlive()) {
            ac->update(dt);
        }
    }
}

void RadarScene::cleanupDestroyedAircraft()
{
    aircraft_.erase(
        std::remove_if(aircraft_.begin(), aircraft_.end(),
            [](const std::unique_ptr<Aircraft>& ac) {
                return !ac->isAlive() ||
                       ac->getRange() > GameConstants::RADAR_MAX_RANGE_KM * 1.1f;
            }),
        aircraft_.end());
}

void RadarScene::runGameLoop()
{
    std::cout << "=== AN/TSQ-73 RADAR CONSOLE ACTIVE ===" << std::endl;
    std::cout << "Running 60-second simulation..." << std::endl;
    std::cout << std::endl;

    const float dt = 1.0f;  // 1-second time steps for console output
    const float totalTime = 60.0f;

    for (float t = 0; t < totalTime; t += dt) {
        update(dt);

        // Print track table every 5 seconds
        if ((int)t % 5 == 0 && trackManager_.getActiveTrackCount() > 0) {
            std::cout << "\n--- TRACK TABLE [T+" << (int)t << "s] ---" << std::endl;
            std::cout << std::left
                      << std::setw(8) << "TRACK"
                      << std::setw(10) << "CLASS"
                      << std::setw(8) << "ALT"
                      << std::setw(8) << "AZM"
                      << std::setw(10) << "RNG(NM)"
                      << std::setw(10) << "SPD(kts)"
                      << std::setw(8) << "HDG"
                      << std::endl;

            auto tracks = trackManager_.getAllTracks();
            for (const auto& track : tracks) {
                std::cout << std::left
                          << std::setw(8) << track.getTrackIdString()
                          << std::setw(10) << track.getClassificationString()
                          << std::setw(8) << track.getAltitudeString()
                          << std::setw(8) << (int)track.azimuth
                          << std::setw(10) << std::fixed << std::setprecision(1)
                          << (track.range * GameConstants::KM_TO_NM)
                          << std::setw(10) << (int)track.speed
                          << std::setw(8) << (int)track.heading
                          << std::endl;
            }
        }

        // Small delay for readable console output
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    // Final battery status
    std::cout << "\n--- BATTERY STATUS ---" << std::endl;
    auto batteries = fireControlSystem_.getAllBatteryData();
    for (const auto& bat : batteries) {
        std::cout << bat.designation << " | Missiles: "
                  << bat.missilesRemaining << "/" << bat.maxMissiles
                  << " | Status: READY" << std::endl;
    }

    std::cout << "\n=== SIMULATION COMPLETE ===" << std::endl;
    std::cout << "Final Score: " << score_ << std::endl;
    std::cout << "Tracks processed: " << trackManager_.getActiveTrackCount()
              << " active" << std::endl;
}
#endif

File: cocos-creator-project/assets/scripts/components/RadarScene.ts
import { _decorator, Component, Node, EventTouch, EventKeyboard, KeyCode, input, Input } from 'cc';
import { GameConstants, EngagementResult, getTrackIdString } from '../game/GameTypes';
import { Aircraft } from '../game/Aircraft';
import { AircraftGenerator } from '../game/AircraftGenerator';
import { TrackManager } from '../game/TrackManager';
import { FireControlSystem } from '../game/FireControlSystem';
import { GameConfig } from '../game/GameConfig';
import { ThreatBoard } from '../game/ThreatBoard';
import { BattalionHQ } from '../game/BattalionHQ';
import { RadarDisplay } from './RadarDisplay';
import { GameHUD } from './GameHUD';

const { ccclass, property } = _decorator;

// ============================================================================
// RadarScene - Main game scene controller (AN/TSQ-73 console)
// ============================================================================

@ccclass('RadarScene')
export class RadarScene extends Component {
    @property(RadarDisplay)
    radarDisplay: RadarDisplay | null = null;

    @property(GameHUD)
    gameHUD: GameHUD | null = null;

    // Core game systems
    private aircraftGenerator = new AircraftGenerator();
    private trackManager = new TrackManager();
    private fireControlSystem = new FireControlSystem();
    private threatBoard = new ThreatBoard();
    private battalionHQ = new BattalionHQ();
    private gameConfig = GameConfig.getInstance();

    // Game state
    private score: number = 0;
    private gameOver: boolean = false;
    private aircraft: Aircraft[] = [];

    start(): void {
        // Initialize game config
        this.gameConfig.setLevel(1);

        // Initialize subsystems
        this.aircraftGenerator.init(this.gameConfig.getLevel());
        this.fireControlSystem.init();
        this.battalionHQ.init(0.5, 0.0);
        this.trackManager.getIFFSystem().setErrorRate(this.gameConfig.getIFFErrorRate());

        // Connect data sources to display components
        if (this.radarDisplay) {
            this.radarDisplay.setTrackManager(this.trackManager);
            this.radarDisplay.setFireControlSystem(this.fireControlSystem);
        }

        if (this.gameHUD) {
            this.gameHUD.setTrackManager(this.trackManager);
            this.gameHUD.setFireControlSystem(this.fireControlSystem);
            this.gameHUD.setThreatBoard(this.threatBoard);
            this.gameHUD.setBattalionHQ(this.battalionHQ);
            this.gameHUD.setLevel(this.gameConfig.getLevel());
        }

        // Input handlers
        this.initInputHandlers();
    }

    private initInputHandlers(): void {
        // Touch/click on radar to select tracks
        if (this.radarDisplay) {
            this.radarDisplay.node.on(Node.EventType.TOUCH_END, this.onRadarTouch, this);
        }

        // Keyboard for battery assignment and fire control
        input.on(Input.EventType.KEY_DOWN, this.onKeyDown, this);
    }

    private onRadarTouch(event: EventTouch): void {
        if (!this.radarDisplay) return;

        const uiPos = event.getUILocation();
        const nodePos = this.radarDisplay.node.getComponent('cc.UITransform')!
            .convertToNodeSpaceAR(new (require('cc').Vec3)(uiPos.x, uiPos.y, 0));

        const nearestTrackId = this.radarDisplay.findNearestTrack(nodePos.x, nodePos.y);
        this.onTrackSelected(nearestTrackId);
    }

    private onKeyDown(event: EventKeyboard): void {
        const batteryMap: Record<number, string> = {
            [KeyCode.DIGIT_1]: 'PATRIOT-1',
            [KeyCode.DIGIT_2]: 'PATRIOT-2',
            [KeyCode.DIGIT_3]: 'PATRIOT-3',
            [KeyCode.DIGIT_4]: 'HAWK-1',
            [KeyCode.DIGIT_5]: 'HAWK-2',
            [KeyCode.DIGIT_6]: 'HAWK-3',
            [KeyCode.DIGIT_7]: 'JAVELIN-1',
            [KeyCode.DIGIT_8]: 'JAVELIN-2',
            [KeyCode.DIGIT_9]: 'JAVELIN-3',
        };

        if (batteryMap[event.keyCode]) {
            this.onBatteryAssign(batteryMap[event.keyCode]);
        } else if (event.keyCode === KeyCode.KEY_F) {
            this.onFireAuthorized();
        } else if (event.keyCode === KeyCode.KEY_A) {
            this.onAbortEngagement();
        }
    }

    update(dt: number): void {
        if (this.gameOver) return;

        this.spawnAircraft(dt);
        this.updateAircraft(dt);
        this.trackManager.update(dt);
        this.fireControlSystem.update(dt);
        this.threatBoard.update(this.trackManager);
        this.battalionHQ.update(dt);
        this.checkEngagements();
        this.cleanupDestroyedAircraft();
        this.checkGameOver();
    }

    private spawnAircraft(dt: number): void {
        const currentCount = this.aircraft.filter((ac) => ac.isAlive()).length;
        const newAircraft = this.aircraftGenerator.trySpawn(dt, currentCount);

        if (newAircraft) {
            this.trackManager.addTrack(newAircraft);
            this.aircraft.push(newAircraft);

            const data = newAircraft.getTrackData();
            const msg = `NEW CONTACT: ${getTrackIdString(data.trackId)} ` +
                        `AZ:${Math.floor(data.azimuth)} RNG:${Math.floor(data.range)}km`;
            if (this.gameHUD) this.gameHUD.addMessage(msg);
        }
    }

    private updateAircraft(dt: number): void {
        for (const ac of this.aircraft) {
            if (ac.isAlive()) {
                ac.update(dt);

                if (ac.hasReachedTerritory() && !ac.isFriendly()) {
                    this.onHostilePenetrated(ac);
                    ac.destroy();
                }
            }
        }
    }

    private checkEngagements(): void {
        const results = this.fireControlSystem.getRecentResults();
        for (const result of results) {
            const aircraft = this.trackManager.getAircraftByTrackId(result.trackId);
            let msg = '';

            if (result.result === EngagementResult.HIT) {
                msg = `${result.batteryDesignation} -> ${aircraft ? getTrackIdString(result.trackId) : 'TK-???'} ** SPLASH **`;
                if (aircraft) {
                    if (aircraft.isFriendly()) {
                        this.onFriendlyDestroyed(aircraft);
                    } else {
                        this.onHostileDestroyed(aircraft);
                    }
                }
            } else if (result.result === EngagementResult.MISS) {
                msg = `${result.batteryDesignation} -> MISS`;
                this.addScore(GameConstants.SCORE_MISSILE_WASTED);
            }

            if (msg && this.gameHUD) {
                this.gameHUD.addMessage(msg);
            }
        }
    }

    private cleanupDestroyedAircraft(): void {
        this.aircraft = this.aircraft.filter((ac) =>
            ac.isAlive() && ac.getRange() <= GameConstants.RADAR_MAX_RANGE_KM * 1.1);
    }

    private checkGameOver(): void {
        if (this.score < -2000) {
            this.gameOver = true;
            if (this.gameHUD) {
                this.gameHUD.addMessage('=== GAME OVER - DEFENSE FAILED ===');
            }
        }
    }

    // Input handlers
    private onTrackSelected(trackId: number): void {
        if (this.radarDisplay) this.radarDisplay.setSelectedTrack(trackId);
        if (this.gameHUD) this.gameHUD.setSelectedTrack(trackId);
    }

    private onBatteryAssign(batteryDesignation: string): void {
        if (!this.radarDisplay || this.radarDisplay.getSelectedTrack() < 0) return;

        const battery = this.fireControlSystem.getBattery(batteryDesignation);
        const aircraft = this.trackManager.getAircraftByTrackId(
            this.radarDisplay.getSelectedTrack());

        if (battery && aircraft && battery.canEngage(aircraft)) {
            if (battery.engage(aircraft)) {
                const msg = `${batteryDesignation} ENGAGING ${getTrackIdString(aircraft.getTrackId())}`;
                if (this.gameHUD) this.gameHUD.addMessage(msg);
            }
        }
    }

    private onFireAuthorized(): void {}
    private onAbortEngagement(): void {}

    // Scoring
    private addScore(points: number): void {
        this.score += points;
        if (this.gameHUD) this.gameHUD.setScore(this.score);
    }

    private onHostileDestroyed(aircraft: Aircraft): void {
        const score = this.gameConfig.getHostileDestroyedScore(aircraft.getType());
        this.addScore(score);
    }

    private onFriendlyDestroyed(_aircraft: Aircraft): void {
        this.addScore(GameConstants.SCORE_FRIENDLY_DESTROYED);
        if (this.gameHUD) {
            this.gameHUD.addMessage('*** WARNING: FRIENDLY FIRE ***');
        }
    }

    private onHostilePenetrated(_aircraft: Aircraft): void {
        this.addScore(GameConstants.SCORE_HOSTILE_PENETRATED);
        if (this.gameHUD) {
            this.gameHUD.addMessage('ALERT: HOSTILE PENETRATED DEFENSE ZONE');
        }
    }

    onDestroy(): void {
        input.off(Input.EventType.KEY_DOWN, this.onKeyDown, this);
    }
}
